core/store/migrate/plugins/relayer/evm/migrate.go

```diff
@@ -0,0 +1,174 @@
+		return nil, fmt.Errorf("failed to get session info: %w", err)
+	}
+	id := fmt.Sprintf("%s@%s:%s", s.User, s.DB, mTable)
+	mu.Lock()
+	defer mu.Unlock()
+	if p, ok := providerCache[id]; ok {
+		return p, nil
+	}
+	// TODO: could be another layer of sharing to reuse generated migrations for the same cfg across different dbs
+	// maybe that would be more error prone and not worth it
+	store, err := database.NewStore(goose.DialectPostgres, mTable)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create goose store for table %s: %w", mTable, err)
+	}
+
+	goMigrations := make([]*goose.Migration, 0)
+	up0002, down0002, err := generate0002(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate migration 0002 for cfg %v: %w", cfg, err)
+	}
+	goMigrations = append(goMigrations, goose.NewGoMigration(2, up0002, down0002))
+
+	// note we are leaking here, but can't delete the temp dir until the migrations are actually executed
+	// maybe update the cache to store the temp dir and delete it when cache is deleted
+	tmpDir, err := os.MkdirTemp("", cfg.Schema)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp dir: %w", err)
+	}
+
+	d := filepath.Join(tmpDir, cfg.Schema, cfg.ChainID.String())
+	err = os.MkdirAll(d, os.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create migration directory %s: %w", d, err)
+	}
+	migrations, err := generateMigrations(embeddedTmplFS, MigrationRootDir, d, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate migrations for opt %v: %w", cfg, err)
+	}
+	fmt.Printf("Generated migrations: %v\n", migrations)
+	fsys := os.DirFS(d)
+
+	// hack to avoid global state. the goose lib doesn't allow to disable the global registry
+	// and also pass custom go migrations (wtf the point of disabling the global registry then?)
+	// https://github.com/pressly/goose/blob/3ad137847a4c242f09e425a12c15c7c7961d4b0f/provider.go#L119
+	goose.ResetGlobalMigrations()
+	p, err := goose.NewProvider(
+		"",
+		db.DB, fsys,
+		goose.WithStore(store),
+		goose.WithDisableGlobalRegistry(true), // until/if we refactor the core migrations to use goose provider
+		goose.WithGoMigrations(goMigrations...))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create goose provider: %w", err)
+	}
+	providerCache[id] = p
+	return p, nil
+}
+
+/*
+func setupPluginMigrations(cfg Cfg) error {
+	// reset the base fs and the global migrations
+	goose.SetBaseFS(nil) // we don't want to use the base fs for plugin migrations because the embedded fs contains templates, not sql files
+	goose.ResetGlobalMigrations()
+	goose.SetTableName(fmt.Sprintf("goose_migration_relayer_%s_%s", cfg.Schema, cfg.ChainID.String()))
+	err := Register0002(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to register migration 0002: %w", err)
+	}
+	return nil
+}
+*/
+// Migrate migrates a subsystem of the chainlink database.
+// It generates migrations based on the template for the subsystem and applies them to the database.
```

Comment:
TODO: update comment to be evm specific (I assume "chainlink database" will no longer be a meaningful concept, right?)